src/plugins/media/media.js

```diff
@@ -0,0 +1,172 @@
+ */
+/* global window, document, impress, console */
+
+(function (document, window) {
+  "use strict";
+  
+  // function names
+  var parseMediaShortcut, 
+      enhanceMediaNode, 
+      enhanceMedia, 
+      playOnStepenter,
+      pauseOnStepleave, 
+      onPlay, 
+      onPause, 
+      onEnded;
+
+  // Parse nodes with media class <div class="media" data-media-source="PATH" data-media-type="MIMETYPE"></div>
+  parseMediaShortcut = function (target) {
+    var data, mimeParts, controls, playonenter, pauseonleave;
+    data = target.dataset;
+    if (data.mediaType && data.mediaSource) {
+      mimeParts = data.mediaType.split('/');
+      target.classList.add('media-' + mimeParts[0]);
+      controls = data.mediaControls === "false" ? '' : ' controls';
+      playonenter = data.mediaPlayonenter === "true" || data.mediaPlayonenter === "" ? ' data-media-playonenter="true"' : '';
+      pauseonleave = data.mediaPauseonleave === "false" ? ' data-media-pauseonleave="false"' : '';
+      target.innerHTML = '<' + mimeParts[0] + controls + playonenter + pauseonleave + '><source src="' + data.mediaSource + '" type="' + data.mediaType + '"> Your Browser does not support HTML5 ' + mimeParts[0] + '.</' + mimeParts[0] + '>';
+    }
+  };
+  
+  onPlay = function (event) {
+    var type = event.target.nodeName.toLowerCase();
+    document.body.classList.add(type + "playing");
+    document.body.classList.remove(type + "paused");
+  };
+  
+  onPause = function (event) {
+    var type = event.target.nodeName.toLowerCase();
+    document.body.classList.add(type + "paused");
+    document.body.classList.remove(type + "playing");
+  };
+  
+  onEnded = function (event) {
+    var type = event.target.nodeName.toLowerCase();
+    document.body.classList.remove(type + "playing");
+    document.body.classList.remove(type + "paused");
+  };
+  
+  enhanceMediaNode = function (type) {
+    var i, id, media = document.getElementsByTagName(type);
+    for (i = 0; i < media.length; i++) {
+      // Set an id to identify each media node - used e.g. by the consoleMedia plugin
+      id = media[i].getAttribute('id');
+      if (id === undefined || id === null) {
+        media[i].setAttribute('id', 'media-' + type + '-' + i);
+      }
+      media[i].addEventListener("play", onPlay, false);
+      media[i].addEventListener("playing", onPlay, false);
+      media[i].addEventListener("pause", onPause, false);
+      media[i].addEventListener("ended", onEnded, false);
+    }
+  };
+  
+  enhanceMedia = function () {
+    var i, steps, media = document.getElementsByClassName('media');
+    for (i = 0; i < media.length; i++) {
+      parseMediaShortcut(media[i]);
+    }
+    enhanceMediaNode('audio');
+    enhanceMediaNode('video');
+    steps = document.getElementsByClassName('step');
+    for (i = 0; i < steps.length; i++) {
+      console.log("Add event listener.");
+      steps[i].addEventListener('impress:stepenter', playOnStepenter, false);
```

Comment:
Ok, here's the hard part...

impress.js now supports a `impress().tear()`. It will reset everything to its original state, as things were before `impress().init()` was called. This means that any changes your plugin does, must be recorded so that they can be reverted when `tear()` is called. This isn't documented in the plugin README, sorry about that.

To add event listeners, use `impress().lib.gc.addEventListener()`. Example: https://github.com/impress/impress.js/blob/master/src/plugins/form/form.js#L33

While on this topic, you must also create and register a callback function, that will remove any classes you added to `body`. Example: https://github.com/impress/impress.js/blob/master/src/plugins/rel/rel.js#L146

(In your case, callback function is simpler, just do a few `classList.remove()` calls and you're done.)

Reply:
What about the defined IDs? (see `enhanceMediaNode()`) Should the be removed as well? How do I record which IDs were added and distinguish them from already existent IDs? Of course one could use a naming convention, but what if some user sets a id that matches that convention accidently? It would be removed as well.

Reply:
Yes. Anything you create should also be removed.

Unfortunately there isn't an easy `gc` helper for removing attributes. You need to create your own function and add it to the gc stack with `gc.pushCallback()`. (Or add the needed helper function into `src/lib/gc.js` if you want!)

    // Set an id to identify each media node - used e.g. by the consoleMedia plugin
    mediaElement = media[i];
    id = mediaElement.getAttribute('id');
    if (id === undefined || id === null) {
      mediaElement.setAttribute('id', 'media-' + type + '-' + i);
      gc.pushCallback( function() { mediaElement.removeAttribute('id'); } );
    }

Reply:
Adding a new helper function into `src/lib/gc.js` is maybe an idea for a later enhancement of the lib. For the moment your suggestion is enough, to work, thank you.